File: src/stores/wiki.ts
import { create } from 'zustand';
import { toast } from 'sonner';
import { getTransport } from '../lib/transport';

// Types matching the Rust structs
export interface WikiArticle {
  id: string;
  tag_id: string;
  content: string;
  created_at: string;
  updated_at: string;
  atom_count: number;
}

export interface WikiCitation {
  id: string;
  citation_index: number;
  atom_id: string;
  chunk_index: number | null;
  excerpt: string;
}

export interface WikiArticleWithCitations {
  article: WikiArticle;
  citations: WikiCitation[];
}

export interface WikiArticleStatus {
  has_article: boolean;
  article_atom_count: number;
  current_atom_count: number;
  new_atoms_available: number;
  updated_at: string | null;
}

export interface WikiArticleSummary {
  id: string;
  tag_id: string;
  tag_name: string;
  updated_at: string;
  atom_count: number;
  inbound_links: number;
}

export interface WikiLink {
  id: string;
  source_article_id: string;
  target_tag_name: string;
  target_tag_id: string | null;
  has_article: boolean;
}

export interface RelatedTag {
  tag_id: string;
  tag_name: string;
  score: number;
  shared_atoms: number;
  semantic_edges: number;
  has_article: boolean;
}

export interface SuggestedArticle {
  tag_id: string;
  tag_name: string;
  atom_count: number;
  mention_count: number;
  score: number;
}

export interface WikiVersionSummary {
  id: string;
  version_number: number;
  atom_count: number;
  created_at: string;
}

export interface WikiArticleVersion {
  id: string;
  tag_id: string;
  content: string;
  citations: WikiCitation[];
  atom_count: number;
  version_number: number;
  created_at: string;
}

export type WikiSectionOp =
  | { op: 'NoChange' }
  | { op: 'AppendToSection'; heading: string; content: string }
  | { op: 'ReplaceSection'; heading: string; content: string }
  | { op: 'InsertSection'; after_heading: string | null; heading: string; content: string };

export interface WikiProposal {
  id: string;
  tag_id: string;
  base_article_id: string;
  base_updated_at: string;
  content: string;
  citations: WikiCitation[];
  ops: WikiSectionOp[];
  new_atom_count: number;
  created_at: string;
}

type WikiView = 'list' | 'article';

interface WikiStore {
  // View state
  view: WikiView;
  currentTagId: string | null;
  currentTagName: string | null;

  // Articles list state
  articles: WikiArticleSummary[];
  isLoadingList: boolean;

  // Suggestions state
  suggestedArticles: SuggestedArticle[];
  isLoadingSuggestions: boolean;

  // Current article state
  currentArticle: WikiArticleWithCitations | null;
  articleStatus: WikiArticleStatus | null;
  relatedTags: RelatedTag[];
  wikiLinks: WikiLink[];

  // Version history
  versions: WikiVersionSummary[];
  selectedVersion: WikiArticleVersion | null;

  // Proposal state (human-in-the-loop update review)
  proposal: WikiProposal | null;
  isProposing: boolean;
  isAccepting: boolean;
  isDismissing: boolean;
  reviewingProposal: boolean;

  // Loading states
  isLoading: boolean;
  isGenerating: boolean;
  isUpdating: boolean;
  error: string | null;

  // List actions
  fetchAllArticles: () => Promise<void>;
  fetchSuggestedArticles: () => Promise<void>;
  showList: () => void;
  openArticle: (tagId: string, tagName: string) => void;
  openAndGenerate: (tagId: string, tagName: string) => void;
  goBack: () => void;

  // Article actions
  fetchArticle: (tagId: string) => Promise<void>;
  fetchArticleStatus: (tagId: string) => Promise<void>;
  fetchRelatedTags: (tagId: string) => Promise<void>;
  fetchWikiLinks: (tagId: string) => Promise<void>;
  generateArticle: (tagId: string, tagName: string) => Promise<void>;
  updateArticle: (tagId: string, tagName: string) => Promise<void>;
  deleteArticle: (tagId: string) => Promise<void>;
  fetchVersions: (tagId: string) => Promise<void>;
  selectVersion: (versionId: string) => Promise<void>;
  clearSelectedVersion: () => void;
  clearArticle: () => void;
  clearError: () => void;
  reset: () => void;

  // Proposal actions (human-in-the-loop update review)
  fetchProposal: (tagId: string) => Promise<void>;
  proposeArticle: (tagId: string, tagName: string) => Promise<void>;
  acceptProposal: (tagId: string) => Promise<void>;
  dismissProposal: (tagId: string) => Promise<void>;
  startReviewingProposal: () => void;
  stopReviewingProposal: () => void;
}

export const useWikiStore = create<WikiStore>((set, get) => ({
  // View state
  view: 'list',
  currentTagId: null,
  currentTagName: null,

  // Articles list state
  articles: [],
  isLoadingList: false,

  // Suggestions state
  suggestedArticles: [],
  isLoadingSuggestions: false,

  // Current article state
  currentArticle: null,
  articleStatus: null,
  relatedTags: [],
  wikiLinks: [],
  versions: [],
  selectedVersion: null,
  proposal: null,
  isProposing: false,
  isAccepting: false,
  isDismissing: false,
  reviewingProposal: false,
  isLoading: false,
  isGenerating: false,
  isUpdating: false,
  error: null,

  fetchAllArticles: async () => {
    set({ isLoadingList: true, error: null });
    try {
      const articles = await getTransport().invoke<WikiArticleSummary[]>('get_all_wiki_articles');
      set({ articles, isLoadingList: false });
      // Refresh suggestions after a brief yield so the list renders first
      setTimeout(() => get().fetchSuggestedArticles(), 50);
    } catch (error) {
      set({ error: String(error), isLoadingList: false });
    }
  },

  fetchSuggestedArticles: async () => {
    set({ isLoadingSuggestions: true });
    try {
      const suggestions = await getTransport().invoke<SuggestedArticle[]>('get_suggested_wiki_articles', { limit: 100 });
      set({ suggestedArticles: suggestions, isLoadingSuggestions: false });
    } catch (error) {
      console.error('Failed to fetch suggested articles:', error);
      toast.error('Failed to load suggested articles', { id: 'wiki-suggestions-error', description: String(error) });
      set({ isLoadingSuggestions: false });
    }
  },

  showList: () => {
    set({
      view: 'list',
      currentTagId: null,
      currentTagName: null,
      currentArticle: null,
      articleStatus: null,
      relatedTags: [],
      wikiLinks: [],
      versions: [],
      selectedVersion: null,
      error: null,
    });
  },

  openArticle: (tagId: string, tagName: string) => {
    set({
      view: 'article',
      currentTagId: tagId,
      currentTagName: tagName,
      currentArticle: null,
      articleStatus: null,
      relatedTags: [],
      wikiLinks: [],
      versions: [],
      selectedVersion: null,
      proposal: null,
      reviewingProposal: false,
      isLoading: true,
      error: null,
    });
    // Fetch article, status, related tags, wiki links, versions, and proposal
    get().fetchArticle(tagId);
    get().fetchArticleStatus(tagId);
    get().fetchRelatedTags(tagId);
    get().fetchWikiLinks(tagId);
    get().fetchVersions(tagId);
    get().fetchProposal(tagId);
  },

  // Open article view and immediately start generating (for new wikis)
  openAndGenerate: (tagId: string, tagName: string) => {
    set({
      view: 'article',
      currentTagId: tagId,
      currentTagName: tagName,
      currentArticle: null,
      articleStatus: null,
      relatedTags: [],
      wikiLinks: [],
      versions: [],
      selectedVersion: null,
      isLoading: false,
      isGenerating: true,
      error: null,
    });
    // Fetch status for display during generation
    get().fetchArticleStatus(tagId);
    // Start generation
    get().generateArticle(tagId, tagName);
  },

  goBack: () => {
    set({
      view: 'list',
      currentTagId: null,
      currentTagName: null,
      currentArticle: null,
      articleStatus: null,
      relatedTags: [],
      wikiLinks: [],
      versions: [],
      selectedVersion: null,
      error: null,
    });
    // Refresh list in case changes were made
    get().fetchAllArticles();
  },

  fetchArticle: async (tagId: string) => {
    set({ isLoading: true, error: null });
    try {
      const article = await getTransport().invoke<WikiArticleWithCitations | null>('get_wiki_article', { tagId });
      set({ currentArticle: article, isLoading: false });
    } catch (error) {
      set({ error: String(error), isLoading: false });
    }
  },

  fetchArticleStatus: async (tagId: string) => {
    try {
      const status = await getTransport().invoke<WikiArticleStatus>('get_wiki_article_status', { tagId });
      set({ articleStatus: status });
    } catch (error) {
      console.error('Failed to fetch article status:', error);
      toast.error('Failed to load article status', { id: 'wiki-status-error', description: String(error) });
    }
  },

  fetchRelatedTags: async (tagId: string) => {
    try {
      const tags = await getTransport().invoke<RelatedTag[]>('get_related_tags', { tagId, limit: 10 });
      set({ relatedTags: tags });
    } catch (error) {
      console.error('Failed to fetch related tags:', error);
      toast.error('Failed to load related tags', { id: 'wiki-related-error', description: String(error) });
    }
  },

  fetchWikiLinks: async (tagId: string) => {
    try {
      const links = await getTransport().invoke<WikiLink[]>('get_wiki_links', { tagId });
      set({ wikiLinks: links });
    } catch (error) {
      console.error('Failed to fetch wiki links:', error);
      toast.error('Failed to load wiki links', { id: 'wiki-links-error', description: String(error) });
    }
  },

  generateArticle: async (tagId: string, tagName: string) => {
    set({
      isGenerating: true,
      error: null,
      selectedVersion: null,
      // Any pending proposal was computed against the article we're about
      // to replace, so it's meaningless now. Backend clears the row; clear
      // the in-memory mirror + exit review mode so the UI stays consistent.
      proposal: null,
      reviewingProposal: false,
    });
    try {
      const article = await getTransport().invoke<WikiArticleWithCitations>('generate_wiki_article', { tagId, tagName });
      set({ currentArticle: article, isGenerating: false });
      // Refresh status, related tags, wiki links, and versions after generation
      get().fetchArticleStatus(tagId);
      get().fetchRelatedTags(tagId);
      get().fetchWikiLinks(tagId);
      get().fetchVersions(tagId);
      // Also refresh the list to include the new article
      get().fetchAllArticles();
    } catch (error) {
      set({ error: String(error), isGenerating: false });
    }
  },

  updateArticle: async (tagId: string, tagName: string) => {
    set({ isUpdating: true, error: null, selectedVersion: null });
    try {
      const article = await getTransport().invoke<WikiArticleWithCitations>('update_wiki_article', { tagId, tagName });
      set({ currentArticle: article, isUpdating: false });
      // Refresh status, related tags, wiki links, and versions after update
      get().fetchArticleStatus(tagId);
      get().fetchRelatedTags(tagId);
      get().fetchWikiLinks(tagId);
      get().fetchVersions(tagId);
      // Also refresh the list
      get().fetchAllArticles();
    } catch (error) {
      set({ error: String(error), isUpdating: false });
    }
  },

  deleteArticle: async (tagId: string) => {
    try {
      await getTransport().invoke('delete_wiki_article', { tagId });
      set({
        currentArticle: null,
        articleStatus: null,
        relatedTags: [],
        wikiLinks: [],
        versions: [],
        selectedVersion: null,
        proposal: null,
        reviewingProposal: false,
      });
      // Refresh the list
      get().fetchAllArticles();
    } catch (error) {
      set({ error: String(error) });
    }
  },

  // ==================== Proposal actions ====================

  fetchProposal: async (tagId: string) => {
    try {
      const proposal = await getTransport().invoke<WikiProposal | null>('get_wiki_proposal', { tagId });
      set({ proposal });
    } catch (error) {
      // 404 is expected when there's no pending proposal — don't toast.
      const msg = String(error);
      if (!msg.includes('No pending proposal')) {
        console.error('Failed to fetch proposal:', error);
      }
      set({ proposal: null });
    }
  },

  proposeArticle: async (tagId: string, tagName: string) => {
    set({ isProposing: true, error: null });
    try {
      const result = await getTransport().invoke<WikiProposal | { status: string }>(
        'propose_wiki_article',
        { tagId, tagName }
      );
      if ('status' in result && result.status === 'no_update_needed') {
        set({ isProposing: false, proposal: null });
        toast.info('No update needed', {
          id: 'wiki-propose-noop',
          description: 'The new atoms don\'t warrant changes to the article.',
        });
        // Refresh status so the banner reflects the latest atom count.
        get().fetchArticleStatus(tagId);
      } else {
        set({ proposal: result as WikiProposal, isProposing: false });
        toast.success('Suggested update ready', {
          id: 'wiki-proposal-ready',
          description: 'Click Review to see what would change.',
        });
      }
    } catch (error) {
      set({ error: String(error), isProposing: false });
      toast.error('Failed to generate update', {
        id: 'wiki-propose-error',
        description: String(error),
      });
    }
  },

  acceptProposal: async (tagId: string) => {
    set({ isAccepting: true, error: null });
    try {
      const article = await getTransport().invoke<WikiArticleWithCitations>(
        'accept_wiki_proposal',
        { tagId }
      );
      set({
        currentArticle: article,
        proposal: null,
        reviewingProposal: false,
        isAccepting: false,
      });
      // Refresh dependent state after the live article changes.
      get().fetchArticleStatus(tagId);
      get().fetchRelatedTags(tagId);
      get().fetchWikiLinks(tagId);
      get().fetchVersions(tagId);
      get().fetchAllArticles();
      toast.success('Update applied');
    } catch (error) {
      const msg = String(error);
      set({ isAccepting: false });
      if (msg.includes('stale')) {
        // Proposal was superseded / live article updated out-of-band.
        set({ proposal: null, reviewingProposal: false });
        toast.error('Proposal out of date', {
          id: 'wiki-proposal-stale',
          description: 'The article was updated elsewhere. Regenerate the suggestion to review it.',
        });
        get().fetchArticle(tagId);
        get().fetchArticleStatus(tagId);
      } else {
        set({ error: msg });
        toast.error('Failed to apply update', {
          id: 'wiki-accept-error',
          description: msg,
        });
      }
    }
  },

  dismissProposal: async (tagId: string) => {
    set({ isDismissing: true });
    try {
      await getTransport().invoke('dismiss_wiki_proposal', { tagId });
      set({ proposal: null, reviewingProposal: false, isDismissing: false });
      get().fetchArticleStatus(tagId);
    } catch (error) {
      set({ isDismissing: false });
      toast.error('Failed to dismiss suggestion', {
        id: 'wiki-dismiss-error',
        description: String(error),
      });
    }
  },

  startReviewingProposal: () => {
    set({ reviewingProposal: true });
  },

  stopReviewingProposal: () => {
    set({ reviewingProposal: false });
  },

  fetchVersions: async (tagId: string) => {
    try {
      const versions = await getTransport().invoke<WikiVersionSummary[]>('get_wiki_versions', { tagId });
      set({ versions });
    } catch (error) {
      console.error('Failed to fetch wiki versions:', error);
      toast.error('Failed to load version history', { id: 'wiki-versions-error', description: String(error) });
    }
  },

  selectVersion: async (versionId: string) => {
    try {
      const version = await getTransport().invoke<WikiArticleVersion | null>('get_wiki_version', { versionId });
      set({ selectedVersion: version });
    } catch (error) {
      console.error('Failed to fetch wiki version:', error);
      toast.error('Failed to load version', { id: 'wiki-version-error', description: String(error) });
    }
  },

  clearSelectedVersion: () => {
    set({ selectedVersion: null });
  },

  clearArticle: () => {
    set({
      currentArticle: null,
      articleStatus: null,
      relatedTags: [],
      wikiLinks: [],
      versions: [],
      selectedVersion: null,
      proposal: null,
      reviewingProposal: false,
      error: null,
    });
  },

  clearError: () => {
    set({ error: null });
  },

  reset: () => {
    set({
      view: 'list',
      currentTagId: null,
      currentTagName: null,
      articles: [],
      isLoadingList: false,
      suggestedArticles: [],
      isLoadingSuggestions: false,
      currentArticle: null,
      articleStatus: null,
      relatedTags: [],
      wikiLinks: [],
      versions: [],
      selectedVersion: null,
      proposal: null,
      isProposing: false,
      isAccepting: false,
      isDismissing: false,
      reviewingProposal: false,
      isLoading: false,
      isGenerating: false,
      isUpdating: false,
      error: null,
    });
  },
}));

File: crates/atomic-server/src/routes/wiki.rs
//! Wiki article routes

use crate::db_extractor::Db;
use crate::error::{ok_or_error, ApiErrorResponse};
use actix_web::{web, HttpResponse};
use serde::{Deserialize, Serialize};
use utoipa::{IntoParams, ToSchema};

#[utoipa::path(get, path = "/api/wiki", responses((status = 200, description = "All wiki articles", body = Vec<atomic_core::WikiArticleSummary>)), tag = "wiki")]
pub async fn get_all_wiki_articles(db: Db) -> HttpResponse {
    ok_or_error(db.0.get_all_wiki_articles().await)
}

#[utoipa::path(get, path = "/api/wiki/{tag_id}", params(("tag_id" = String, Path, description = "Tag ID")), responses((status = 200, description = "Wiki article with citations", body = atomic_core::WikiArticleWithCitations), (status = 404, description = "No article for tag", body = ApiErrorResponse)), tag = "wiki")]
pub async fn get_wiki(db: Db, path: web::Path<String>) -> HttpResponse {
    let tag_id = path.into_inner();
    ok_or_error(db.0.get_wiki(&tag_id).await)
}

#[utoipa::path(get, path = "/api/wiki/{tag_id}/status", params(("tag_id" = String, Path, description = "Tag ID")), responses((status = 200, description = "Wiki article status", body = atomic_core::WikiArticleStatus)), tag = "wiki")]
pub async fn get_wiki_status(db: Db, path: web::Path<String>) -> HttpResponse {
    let tag_id = path.into_inner();
    ok_or_error(db.0.get_wiki_status(&tag_id).await)
}

#[derive(Deserialize, Serialize, ToSchema)]
pub struct GenerateWikiBody {
    /// Tag name for the wiki article
    pub tag_name: String,
}

#[utoipa::path(post, path = "/api/wiki/{tag_id}/generate", params(("tag_id" = String, Path, description = "Tag ID")), request_body = GenerateWikiBody, responses((status = 200, description = "Generated wiki article", body = atomic_core::WikiArticleWithCitations), (status = 400, description = "Error", body = ApiErrorResponse)), tag = "wiki")]
pub async fn generate_wiki(
    db: Db,
    path: web::Path<String>,
    body: web::Json<GenerateWikiBody>,
) -> HttpResponse {
    let tag_id = path.into_inner();
    match db.0.generate_wiki(&tag_id, &body.tag_name).await {
        Ok(article) => HttpResponse::Ok().json(article),
        Err(e) => crate::error::error_response(e),
    }
}

#[utoipa::path(post, path = "/api/wiki/{tag_id}/update", params(("tag_id" = String, Path, description = "Tag ID")), request_body = GenerateWikiBody, responses((status = 200, description = "Updated wiki article", body = atomic_core::WikiArticleWithCitations), (status = 400, description = "Error", body = ApiErrorResponse)), tag = "wiki")]
pub async fn update_wiki(
    db: Db,
    path: web::Path<String>,
    body: web::Json<GenerateWikiBody>,
) -> HttpResponse {
    let tag_id = path.into_inner();
    match db.0.update_wiki(&tag_id, &body.tag_name).await {
        Ok(article) => HttpResponse::Ok().json(article),
        Err(e) => crate::error::error_response(e),
    }
}

#[utoipa::path(delete, path = "/api/wiki/{tag_id}", params(("tag_id" = String, Path, description = "Tag ID")), responses((status = 200, description = "Wiki deleted")), tag = "wiki")]
pub async fn delete_wiki(db: Db, path: web::Path<String>) -> HttpResponse {
    let tag_id = path.into_inner();
    ok_or_error(db.0.delete_wiki(&tag_id).await)
}

#[derive(Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub struct RelatedTagsQuery {
    /// Max results (default: 10)
    pub limit: Option<usize>,
}

#[utoipa::path(get, path = "/api/wiki/{tag_id}/related", params(("tag_id" = String, Path, description = "Tag ID"), RelatedTagsQuery), responses((status = 200, description = "Related tags", body = Vec<atomic_core::RelatedTag>)), tag = "wiki")]
pub async fn get_related_tags(
    db: Db,
    path: web::Path<String>,
    query: web::Query<RelatedTagsQuery>,
) -> HttpResponse {
    let tag_id = path.into_inner();
    let limit = query.limit.unwrap_or(10);
    ok_or_error(db.0.get_related_tags(&tag_id, limit).await)
}

#[utoipa::path(get, path = "/api/wiki/{tag_id}/links", params(("tag_id" = String, Path, description = "Tag ID")), responses((status = 200, description = "Wiki cross-reference links", body = Vec<atomic_core::WikiLink>)), tag = "wiki")]
pub async fn get_wiki_links(db: Db, path: web::Path<String>) -> HttpResponse {
    let tag_id = path.into_inner();
    ok_or_error(db.0.get_wiki_links(&tag_id).await)
}

#[derive(Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub struct SuggestionsQuery {
    /// Max suggestions (default: 10)
    pub limit: Option<i32>,
}

#[utoipa::path(get, path = "/api/wiki/suggestions", params(SuggestionsQuery), responses((status = 200, description = "Suggested wiki articles", body = Vec<atomic_core::SuggestedArticle>)), tag = "wiki")]
pub async fn get_wiki_suggestions(
    db: Db,
    query: web::Query<SuggestionsQuery>,
) -> HttpResponse {
    let limit = query.limit.unwrap_or(10);
    ok_or_error(db.0.get_suggested_wiki_articles(limit).await)
}

#[utoipa::path(get, path = "/api/wiki/{tag_id}/versions", params(("tag_id" = String, Path, description = "Tag ID")), responses((status = 200, description = "Version history", body = Vec<atomic_core::WikiVersionSummary>)), tag = "wiki")]
pub async fn list_wiki_versions(db: Db, path: web::Path<String>) -> HttpResponse {
    let tag_id = path.into_inner();
    ok_or_error(db.0.list_wiki_versions(&tag_id).await)
}

#[utoipa::path(get, path = "/api/wiki/versions/{version_id}", params(("version_id" = String, Path, description = "Version ID")), responses((status = 200, description = "Wiki article version", body = atomic_core::WikiArticleVersion)), tag = "wiki")]
pub async fn get_wiki_version(db: Db, path: web::Path<String>) -> HttpResponse {
    let version_id = path.into_inner();
    ok_or_error(db.0.get_wiki_version(&version_id).await)
}

#[utoipa::path(post, path = "/api/wiki/recompute-tag-embeddings", responses((status = 200, description = "Recomputed tag embeddings")), tag = "wiki")]
pub async fn recompute_all_tag_embeddings(db: Db) -> HttpResponse {
    match db.0.recompute_all_tag_embeddings().await {
        Ok(count) => HttpResponse::Ok().json(serde_json::json!({"count": count})),
        Err(e) => crate::error::error_response(e),
    }
}

// ==================== Wiki Proposals (human-in-the-loop update review) ====================

#[utoipa::path(
    post,
    path = "/api/wiki/{tag_id}/propose",
    params(("tag_id" = String, Path, description = "Tag ID")),
    request_body = GenerateWikiBody,
    responses(
        (status = 200, description = "Proposal created or no update needed", body = atomic_core::WikiProposal),
        (status = 400, description = "Error", body = ApiErrorResponse)
    ),
    tag = "wiki"
)]
pub async fn propose_wiki(
    db: Db,
    path: web::Path<String>,
    body: web::Json<GenerateWikiBody>,
) -> HttpResponse {
    let tag_id = path.into_inner();
    match db.0.propose_wiki_update(&tag_id, &body.tag_name).await {
        Ok(Some(proposal)) => HttpResponse::Ok().json(proposal),
        Ok(None) => HttpResponse::Ok().json(serde_json::json!({
            "status": "no_update_needed"
        })),
        Err(e) => crate::error::error_response(e),
    }
}

#[utoipa::path(
    get,
    path = "/api/wiki/{tag_id}/proposal",
    params(("tag_id" = String, Path, description = "Tag ID")),
    responses(
        (status = 200, description = "Pending wiki proposal", body = atomic_core::WikiProposal),
        (status = 404, description = "No pending proposal", body = ApiErrorResponse)
    ),
    tag = "wiki"
)]
pub async fn get_wiki_proposal(db: Db, path: web::Path<String>) -> HttpResponse {
    let tag_id = path.into_inner();
    match db.0.get_wiki_proposal(&tag_id).await {
        Ok(Some(proposal)) => HttpResponse::Ok().json(proposal),
        Ok(None) => HttpResponse::NotFound()
            .json(serde_json::json!({"error": "No pending proposal for this tag"})),
        Err(e) => crate::error::error_response(e),
    }
}

#[utoipa::path(
    post,
    path = "/api/wiki/{tag_id}/proposal/accept",
    params(("tag_id" = String, Path, description = "Tag ID")),
    responses(
        (status = 200, description = "Proposal accepted and promoted to live article", body = atomic_core::WikiArticleWithCitations),
        (status = 400, description = "Error", body = ApiErrorResponse)
    ),
    tag = "wiki"
)]
pub async fn accept_wiki_proposal(db: Db, path: web::Path<String>) -> HttpResponse {
    let tag_id = path.into_inner();
    match db.0.accept_wiki_proposal(&tag_id).await {
        Ok(article) => HttpResponse::Ok().json(article),
        Err(e) => crate::error::error_response(e),
    }
}

#[utoipa::path(
    post,
    path = "/api/wiki/{tag_id}/proposal/dismiss",
    params(("tag_id" = String, Path, description = "Tag ID")),
    responses(
        (status = 204, description = "Proposal dismissed"),
        (status = 400, description = "Error", body = ApiErrorResponse)
    ),
    tag = "wiki"
)]
pub async fn dismiss_wiki_proposal(db: Db, path: web::Path<String>) -> HttpResponse {
    let tag_id = path.into_inner();
    match db.0.dismiss_wiki_proposal(&tag_id).await {
        Ok(()) => HttpResponse::NoContent().finish(),
        Err(e) => crate::error::error_response(e),
    }
}
